lib/fixed-points.nix

```diff
@@ -107,7 +107,7 @@ rec {
   # Same as `makeExtensible` but the name of the extending attribute is
   # customized.
   makeExtensibleWithCustomName = extenderName: rattrs:
-    fix' rattrs // {
+    fix' (self: rattrs self // {
       ${extenderName} = f: makeExtensibleWithCustomName extenderName (extends f rattrs);
-   };
+    });
```

Comment:
Oh I think I get it. Here's an example that illustrates this: Previously if one wanted to `.extend` `lib` within an `.extend` of a `lib` itself, you'd have to manually do a fixed-point in order to have access to the `.extend`:

```nix
let
  result = lib.extend (final: prev: {
    x = 10;

    nested = result.extend (final: prev: {
      y = 10;
    });
  });
in result
```

Because if you just tried to `final.extends`, that doesn't work:

```nix
lib.extend (final: prev: {
  x = 10;

  nested = final.extend (final: prev: {
    y = 10;
  });
})
```

```
error: attribute 'extend' missing

       at /home/infinisil/src/nixpkgs/test.nix:6:12:

            5|
            6|   nested = final.extend (final: prev: {
             |            ^
            7|     y = 10;
```

This change makes that work however.
```

Reply:
Without this, if you do `lib.evalModules { modules = [({lib, ...}: ...)]; }` then the inner `lib` is not extensible. This is a problem in the context of this PR because of things like https://github.com/nix-community/home-manager/blob/master/modules/lib/stdlib-extended.nix#L7

Reply:
Though actually if the flake used `lib = self.outputs.lib` instead of `lib = final`, then it should also work.

I am just a tiny bit worried regarding this change that it might cause infinite recursions for certain scenarios when it didn't before. Because the `.extend` now goes through the fixed-point, meaning that `rattrs` needs to be evaluated before `.extend` can be accessed

Reply:
And `makeExtensible` is used for much of `pkgs`

Reply:
I'd prefer not to change this definition, but add a new definition instead. That way usages can switch one by one. It seems likely to break existing code.

Also, a cleaner solution would be to pass the "extend" function to `rattrs` so it can choose where to put it. That way, we don't need an `extenderName` arg, and no dynamic attrset. I think you'll want to use a `let` instead of `fix'`, but that's clearer and faster anyway.

Another point in this space is [lib.encapsulate](https://github.com/NixOS/nixpkgs/pull/158781); also a new function. Although maybe `lib` does not need an internal / `public` distinction. Probably?

Reply:
> Though actually if the flake used `lib = self.outputs.lib` instead of `lib = final`, then it should also work.

This wouldn't solve the home-manager issue I mentioned (I checked). The problem isn't that `final` doesn't contain `extend`, but rather that `final.evalModules` (or `self.lib.evalModules`) doesn't make `extend` available in modules.

> I am just a tiny bit worried regarding this change that it might cause infinite recursions for certain scenarios when it didn't before, because the `.extend` now goes through the fixed-point, meaning that `rattrs` needs to be evaluated before `.extend` can be accessed

I really cannot think of a practical scenario where this would be a problem. To use `.extend` you need `rattrs self` to be evaluated up to its keys, which surely would not depend on `.extend` in any reasonable scenario? At least everything in nixpkgs still evaluates.

Reply:
> To use `.extend` you need `rattrs self` to be evaluated up to its keys

Oh that's a good point, my increased strictness argument doesn't work because it was just as strict before.

With that not being a thing, I don't think there's anything to worry anymore with this change.

Reply:
I retract my objection to this change.